crates/svm-abi/decoder/src/decoder.rs

```diff
@@ -70,55 +83,171 @@ impl Decoder {
+    fn decode_u8(&self, cursor: &mut Cursor) -> Result<u8, DecodeError> {
+        let num = self.decode_i8(cursor)? as u8;
+        Ok(num)
+    }
+
+    fn decode_i16(&self, cursor: &mut Cursor) -> Result<i16, DecodeError> {
+        let byte = self.read_byte(cursor)?;
+
+        let nbytes = match byte {
+            layout::I16_1B | layout::U16_1B => 1,
+            layout::I16_2B | layout::U16_2B => 2,
+            _ => unreachable!(),
+        };
+
+        let num = self.read_num(cursor, nbytes)? as i16;
+        Ok(num)
+    }
+
+    fn decode_u16(&self, cursor: &mut Cursor) -> Result<u16, DecodeError> {
+        let num = self.decode_i16(cursor)? as u16;
+        Ok(num)
+    }
+
+    fn decode_i32(&self, cursor: &mut Cursor) -> Result<i32, DecodeError> {
+        let byte = self.read_byte(cursor)?;
+
+        let nbytes = match byte {
+            layout::I32_1B | layout::U32_1B => 1,
+            layout::I32_2B | layout::U32_2B => 2,
+            layout::I32_3B | layout::U32_3B => 3,
+            layout::I32_4B | layout::U32_4B => 4,
+            _ => unreachable!(),
+        };
+
+        let num = self.read_num(cursor, nbytes)? as i32;
+        Ok(num)
     }
 
-    fn decode_pubkey256(&self, cursor: &mut Cursor) -> Result<Value, DecodeError> {
-        decode_fixed_primitive!(self, PubKey256, 32, cursor)
+    fn decode_u32(&self, cursor: &mut Cursor) -> Result<u32, DecodeError> {
+        let num = self.decode_i32(cursor)? as u32;
+        Ok(num)
     }
 
-    fn verify_array(&self, _value: &Value) -> Result<(), DecodeError> {
-        // todo!()
-        Ok(())
+    fn decode_i64(&self, cursor: &mut Cursor) -> Result<i64, DecodeError> {
+        let byte = self.read_byte(cursor)?;
+
+        let nbytes = match byte {
+            layout::I64_1B | layout::U64_1B => 1,
+            layout::I64_2B | layout::U64_2B => 2,
+            layout::I64_3B | layout::U64_3B => 3,
+            layout::I64_4B | layout::U64_4B => 4,
+            layout::I64_5B | layout::U64_5B => 5,
+            layout::I64_6B | layout::U64_6B => 6,
+            layout::I64_7B | layout::U64_7B => 7,
+            layout::I64_8B | layout::U64_8B => 8,
+            _ => unreachable!(),
+        };
+
+        let num = self.read_num(cursor, nbytes)? as i64;
+        Ok(num)
+    }
+
+    fn decode_u64(&self, cursor: &mut Cursor) -> Result<u64, DecodeError> {
+        let num = self.decode_i64(cursor)? as u64;
+        Ok(num)
+    }
+
+    fn decode_array(&self, cursor: &mut Cursor) -> Result<Value, DecodeError> {
+        assert_no_eof!(cursor);
+
+        extern crate std;
+        use std::dbg;
+
+        let byte = self.read_byte(cursor)?;
+        let nitems = match byte {
+            layout::ARR_0 => 0,
+            layout::ARR_1 => 1,
+            layout::ARR_2 => 2,
+            layout::ARR_3 => 3,
+            layout::ARR_4 => 4,
+            layout::ARR_5 => 5,
+            layout::ARR_6 => 6,
+            layout::ARR_0_255 => self.read_byte(cursor)?,
```

Comment:
@moshababo here we read a byte to know what's the length of the array.
We know it's between 6 to 255 (inclusive) - but this byte will tell us the exact length.